src/overlays/actors/ovl_Arrow_Fire/z_arrow_fire.c

```diff
@@ -41,29 +37,267 @@ const ActorInit Arrow_Fire_InitVars = {
+};
 
-extern ColliderQuadInit D_80922230;
+static InitChainEntry sInitChain[] = {
+    ICHAIN_F32(uncullZoneForward, 2000, ICHAIN_STOP),
+};
+
+static s32 sUnused;
 
 // there are uses of D_0E000000.fillRect (appearing as D_0E0002E0) in this file
 extern GfxMasterList D_0E000000;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Arrow_Fire/ArrowFire_SetupAction.s")
+void ArrowFire_SetupAction(ArrowFire* this, ArrowFireActionFunc actionFunc) {
+    this->actionFunc = actionFunc;
+}
+
+void ArrowFire_Init(Actor* thisx, GlobalContext* globalCtx) {
+    ArrowFire* this = THIS;
+
+    Actor_ProcessInitChain(&this->actor, sInitChain);
+    this->radius = 0;
+    this->height = 1.0f;
+    ArrowFire_SetupAction(this, FireArrow_ChargeAndWait);
+    Actor_SetScale(&this->actor, 0.01f);
+    this->alpha = 160;
+    this->timer = 0;
+    this->blueingEffectMagnitude = 0.0f;
+    Collider_InitAndSetQuad(globalCtx, &this->collider1, &this->actor, &sQuadInit);
+    Collider_InitAndSetQuad(globalCtx, &this->collider2, &this->actor, &sQuadInit);
+}
+
+void ArrowFire_Destroy(Actor* thisx, GlobalContext* globalCtx) {
+    ArrowFire* this = THIS;
+
+    func_80115D5C(&globalCtx->state);
+    Collider_DestroyQuad(globalCtx, &this->collider1);
+    Collider_DestroyQuad(globalCtx, &this->collider2);
+}
+
+void FireArrow_ChargeAndWait(ArrowFire* this, GlobalContext* globalCtx) {
+    EnArrow* arrow = (EnArrow*)this->actor.parent;
+
+    if ((arrow == NULL) || (arrow->actor.update == NULL)) {
+        Actor_MarkForDeath(&this->actor);
+        return;
+    }
+    if (this->radius < 10) {
+        this->radius++;
+    }
+
+    this->actor.world.pos = arrow->actor.world.pos;
+    this->actor.shape.rot = arrow->actor.shape.rot;
+
+    func_800B9010(&this->actor, NA_SE_PL_ARROW_CHARGE_FIRE - SFX_FLAG);
+
+    // if arrow has no parent, player has fired the arrow
+    if (arrow->actor.parent == NULL) {
+        this->firedPos = this->actor.world.pos;
+        this->radius = 10;
+        ArrowFire_SetupAction(this, FireArrow_Fly);
+        this->alpha = 255;
+    }
+}
+
+void FireArrow_Lerp(Vec3f* firedPos, Vec3f* pos, f32 scale) {
+    VEC3F_LERPIMPDST(firedPos, firedPos, pos, scale);
+}
+
+void FireArrow_Hit(ArrowFire* this, GlobalContext* globalCtx) {
+    f32 offset;
+    f32 scale;
+    u16 timer;
+
+    if (this->actor.projectedW < 50.0f) {
+        scale = 10.0f;
+    } else if (this->actor.projectedW > 950.0f) {
+        scale = 310.0f;
+    } else {
+        scale = this->actor.projectedW;
+        scale = ((scale - 50.0f) * (1.0f / 3.0f)) + 10.0f;
+    }
+
+    timer = this->timer;
+    if (timer != 0) {
+        this->timer--;
+        if (this->timer >= 8) {
+            offset = (this->timer - 8) * (1.0f / 24.0f);
+            offset = SQ(offset);
+            this->radius = (((1.0f - offset) * scale) + 10.0f);
+            this->height = this->height + ((2.0f - this->height) * 0.1f);
```

Comment:
```suggestion
            this->height += (2.0f - this->height) * 0.1f;
```

Reply:
```suggestion
            this->height = F32_LERPIMP(this->height, 2.0f, 0.1f);
```
Sorry actually, I think this would be better (would apply to light arrows as well)

Reply:
I applied this to ice arrows as well.